### DISCUSSION TITLE
How do I work with non primitive types?

### DISCUSSION THREAD

**Discussion by cesalberca** [react-hook-form/react-hook-form#7606] [Q&A]
I'm very happy with this library, so great work!

I have a problem when dealing with a form with non-primitive types and I'm wondering if I can handle them somehow.

I have a custom input that looks like this:

```tsx
import { FC } from 'react'
import { Datetime } from '@archimedes/utils'
import { BaseInput } from '../base-input/base-input'
import { Input } from '../input'

export const InputDate: FC<Input<Datetime>> = ({ name, className, errors, label, value, onChange }) => {
  return (
    <BaseInput name={name} className={className} errors={errors} label={label}>
      <input
        type="date"
        value={value?.format('yyyy-MM-dd') ?? ''}
        onChange={e => onChange(Datetime.fromIsoString(e.target.value))}
      />
    </BaseInput>
  )
}
```

Datetime is a wrapper around Luxon's DateTime and when dealing with forms I would like to handle all dates with this wrapper.

I use react-hook-form as follows:

```tsx
export function FormInputDate<FieldValues>({
  control,
  name,
  className,
  label,
  defaultValue,
  rules,
  shouldUnregister,
}: FormInput<FieldValues>) {
  const { field, formState } = useController({ control, name, rules, shouldUnregister, defaultValue })
  return (
    <InputDate
      name={name}
      value={field.value as Datetime}
      onChange={(x) => field.onChange(x)}
      className={className}
      label={label}
      errors={formState.errors}
    />
  )
}
```

And when I use the `FormInputDate`:

```tsx
<FormInputDate<FormModel> name={'foo'} label={t('bar')} control={control} />
```

The problem is that I get an error, telling me that `value?.format('yyyy-MM-dd')` is not a function.

Is this possible using this library?

**Comment by Moshyfawn** [comment]
You sure there's supposed to be a `format` method on DateTime instance?
<img width="479" alt="image" src="https://user-images.githubusercontent.com/16290753/150356575-f214ff8f-583f-41bd-b47a-85d188f13cdf.png">

**Reply by Moshyfawn** [reply]
Huh
<img width="456" alt="image" src="https://user-images.githubusercontent.com/16290753/150358811-29a50e7b-316c-4be2-a90e-d71023bcc41b.png">

**Reply by cesalberca** [reply]
Yes, since it's not Luxon's DateTime, I have my own wrapper around it and it has a format method. https://github.com/archimedes-projects/archimedes-js/blob/be4e15083f9b84188b58ba8f27e878f1523cf717/packages/utils/src/datetime/datetime.ts

It seems there is some kind of transformation done underneath that breaks the class.

**Comment by Moshyfawn** [comment]
Might be related #3877

**Reply by cesalberca** [reply]
I've created a Sandbox here: https://codesandbox.io/s/react-hook-form-class-breakage-hwjg7?file=/src/input-date.tsx. The issue might be related.

**Reply by Moshyfawn** [reply]
@bluebill1049 it looks like the issue with class instances being mangled by RHF in #3877 issue. [In this comment](https://github.com/react-hook-form/react-hook-form/issues/3877#issuecomment-760547844) you mentioned fixing it in V7. Is it a different issue? Does it need it's own GH issue created?

**Reply by bluebill1049** [reply]
yea this is a design limitation with those custom objects... this feature request is probably what you after: https://github.com/react-hook-form/react-hook-form/issues/7279

**Reply by cesalberca** [reply]
Right! But it's closed, is this something react-hook-forms might provide in the future?

**Reply by bluebill1049** [reply]
It's not closed. It's moved into the project board.
![Screen Shot 2022-01-28 at 7 56 05 pm](https://user-images.githubusercontent.com/10513364/151516892-1c4f586c-af51-4cee-9145-d4b16539eca1.png)

**Reply by tleed5** [reply]
@bluebill1049 
Is this feature still being worked on? I've run into this issue as well with Luxon's datetime types. On form submission in a test environment(jest/jsdom) the luxon datetime loses all of it's class methods where as in the browser it keeps all of it's class details.

Is my only solution at the moment to just convert to and from strings when necessary?
